sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -24,12 +26,34 @@ import scala.util.Try
 import org.apache.spark.annotation.Experimental
 import org.apache.spark.sql.catalyst.{SqlParser, ScalaReflection}
 import org.apache.spark.sql.catalyst.analysis.{UnresolvedFunction, Star}
+import org.apache.spark.sql.catalyst.encoders.{ExpressionEncoder, Encoder}
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.plans.logical.BroadcastHint
+import org.apache.spark.sql.execution.aggregate.SumOf
 import org.apache.spark.sql.types._
 import org.apache.spark.util.Utils
 
 /**
+ * Ensures that java functions signatures for methods that now return a [[TypedColumn]] still have
+ * legacy equivalents in bytecode.  This compatibility is done by forcing the compiler to generate
+ * "bridge" methods due to the use of covariant return types.
+ *
+ * {{{
+ * In LegacyFunctions:
+ * public abstract org.apache.spark.sql.Column avg(java.lang.String);
+ *
+ * In functions:
+ * public static org.apache.spark.sql.TypedColumn<java.lang.Object, java.lang.Object> avg(...);
+ * }}}
+ *
+ * This allows us to use the same functions both in typed [[Dataset]] operations and untyped
+ * [[DataFrame]] operations when the return type for a given function is statically known.
+ */
+abstract class LegacyFunctions {
```

Comment:
@mateiz, did you see this part too?  This will let users call things like `count("*")` without having to manually say `count("*").as[Long]`, which seems pretty nice.  We can also do it without breaking binary compatibility.

Questions:
 - should we do this at all (my vote is yes, at least for very common things like count)`
 - should we do it for all things that return static types
 - should we do it even for things that don't return static types.  (i.e. `avg` returns `Double` or `Decimal`.  we could default to `Double` and require the user to say `.as[BigDecimal]` if they want to maintain precision.